ipa-core/src/protocol/ipa_prf/malicious_security/quadratic_proofs.rs

```diff
@@ -0,0 +1,933 @@
+use futures::StreamExt;
+use futures_util::{stream, TryStreamExt};
+use ipa_macros::Step;
+
+use crate::{
+    error::Error,
+    ff::{Field, Invert},
+    helpers::{Direction, ReceivingEnd, SendingEnd},
+    protocol::{
+        basics::validate_replicated_shares::validate_sum_to_zero,
+        context::Context,
+        ipa_prf::malicious_security::{
+            fiat_shamir::{fiat_shamir_prover, fiat_shamir_verifier},
+            lagrange::{
+                compute_lagrange_base, compute_lagrange_denominator, generate_evaluation_points,
+                lagrange_evaluation, lagrange_evaluation_precomputed,
+            },
+        },
+        prss::SharedRandomness,
+        RecordId,
+    },
+    seq_join::seq_join,
+};
+
+#[derive(Step)]
+pub(crate) enum Step {
+    HashFromLeft,
+    HashFromRight,
+    PQFromLeft,
+    TwoOutOfTwoZeroCheck,
+}
+
+/// Non-Interactive Distributed Zero Knowledge Proof
+/// a `NIDZKP` which is computed over multiple recursions , is a vector of `g` polynomials
+/// the length is the amount of proof recursions
+/// `g` is secret shared using PRSS: left share is stored in struct, right share is generated via PRSS
+/// the length of the left share, `g_left`, determines the degree of the polynomial
+/// for a low communication rate, i.e. `log (amount of multiplications)`,
+/// the length needs to be constant in the amount of multiplications
+/// length might be different across recursions (therefore used `Vec` instead of `GenericArray`)
+/// the "left proof" also contains non-zero masks, `mask_p` and `mask_q`.
+/// For the "right proof" both masks are zero.
+///
+/// Proofs that are sent to other parties have mask that are always zero and therefore don't need to be sent.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct NIDZKP<F>
+where
+    F: Field,
+{
+    pub(crate) proofs: Vec<Vec<F>>,
+    pub(crate) mask_p: F,
+    pub(crate) mask_q: F,
+}
+
+/// generates recursive proof for `sum_i u_i*v_i = out` over field `F`
+///
+/// `r_f`, i.e. recursion factor, determines how much the inputs are compressed during each recursion
```

Comment:
Since `r` usually is associated with the random point, and the paper we are following calls this either `lambda` or the `compression_factor`, can we use one of those terms?